libtsuba/src/ParquetReader.cpp

```diff
@@ -148,141 +120,251 @@ tsuba::ParquetReader::ReadFromUriSliced(const katana::Uri& uri) {
 
-  auto reader_res =
-      MakeFileReader(uri, 0, std::numeric_limits<uint64_t>::max());
-  if (!reader_res) {
-    return reader_res.error();
+    std::unique_ptr<BlockedParquetReader> bpr(new BlockedParquetReader(
+        uri.string(), std::move(fvs), std::move(readers),
+        std::move(row_offsets)));
+
+    if (preload) {
+      for (size_t i = 0, num_files = bpr->row_offsets_.size(); i < num_files;
+           ++i) {
+        KATANA_CHECKED(bpr->EnsureReader(i, true));
+      }
+    }
+    return std::unique_ptr<BlockedParquetReader>(std::move(bpr));
   }
-  std::unique_ptr<parquet::arrow::FileReader> reader(
-      std::move(reader_res.value()));
 
-  std::shared_ptr<arrow::Table> out;
-  auto read_result = reader->ReadTable(&out);
-  if (!read_result.ok()) {
-    return KATANA_ERROR(ErrorCode::ArrowError, "arrow error: {}", read_result);
+  Result<int64_t> NumRows() {
+    size_t last_reader_idx = readers_.size() - 1;
+    KATANA_CHECKED(EnsureReader(last_reader_idx));
+    return row_offsets_[last_reader_idx] +
+           readers_[last_reader_idx]->parquet_reader()->metadata()->num_rows();
   }
-  return FixTable(std::move(out));
-}
 
-Result<std::shared_ptr<arrow::Table>>
-tsuba::ParquetReader::DoFilteredTableRead(
-    parquet::arrow::FileReader* reader, const arrow::Schema& schema,
-    const std::vector<int32_t>& indexes) {
-  if (slice_) {
-    // TODO(thunt) implement this
-    return KATANA_ERROR(
-        ErrorCode::NotImplemented,
-        "sorry! missing support for sliced read when choosing columns");
+  Result<int32_t> NumColumns() {
+    KATANA_CHECKED(EnsureReader(0));
+    return readers_[0]->parquet_reader()->metadata()->num_columns();
   }
 
-  std::vector<std::shared_ptr<arrow::Field>> fields;
-  std::vector<std::shared_ptr<arrow::ChunkedArray>> columns;
-  std::unordered_map<int32_t, std::shared_ptr<arrow::ChunkedArray>> read_arrays;
-  for (int32_t idx : indexes) {
-    if (idx < 0) {
+  Result<std::shared_ptr<arrow::Table>> ReadTable(
+      std::optional<tsuba::ParquetReader::Slice> slice = std::nullopt) {
+    if (!slice) {
+      std::vector<std::shared_ptr<arrow::Table>> tables;
+      for (size_t i = 0, num_files = readers_.size(); i < num_files; ++i) {
+        KATANA_CHECKED(EnsureReader(i, true));
+        std::shared_ptr<arrow::Table> table;
+        KATANA_CHECKED(readers_[i]->ReadTable(&table));
+        tables.emplace_back(std::move(table));
+      }
+      return KATANA_CHECKED(arrow::ConcatenateTables(tables));
+    }
+
+    int64_t curr_global_row = slice->offset;
+    int64_t last_global_row =
+        std::min(KATANA_CHECKED(NumRows()), curr_global_row + slice->length);
+
+    if (last_global_row <= curr_global_row) {
       return KATANA_ERROR(
-          ErrorCode::InvalidArgument, "column indexes must be positive");
+          katana::ErrorCode::InvalidArgument,
+          "slice would result in empty table");
     }
-    if (idx >= schema.num_fields()) {
+
+    auto it = std::upper_bound(
```

Comment:
Are row_offsets unique? If so, is it possible to use `lower_bound` and avoid `it - 1` below? Or equivalently, how do we know that `it` is not `begin`?